force_bdss/app/base_operation.py

```diff
@@ -0,0 +1,48 @@
+from threading import Event as ThreadingEvent
+import logging
+
+from traits.api import (
+    DelegatesTo,
+    HasStrictTraits,
+    Instance,
+    provides
+)
+
+from .i_operation import IOperation
+from .workflow_file import WorkflowFile
+
+log = logging.getLogger(__name__)
+
+
+@provides(IOperation)
+class BaseOperation(HasStrictTraits):
+    """ Base class for EvaluateOperation and OptimizeOperation.
+    """
+
+    #: The workflow file being operated on.
+    workflow_file = Instance(WorkflowFile)
+
+    #: The workflow instance.
+    workflow = DelegatesTo("workflow_file")
+
+    #: Threading Event instance that indicates if the optimization operation
+    #: should be stopped.
+    _stop_event = Instance(ThreadingEvent, visible=False, transient=True)
+
+    #: Threading Event instance that indicates if the optimization operation
+    #: should be paused and then resumed.
+    _pause_event = Instance(ThreadingEvent, visible=False, transient=True)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._stop_event = ThreadingEvent()
+        self._pause_event = ThreadingEvent()
+        self._pause_event.set()
+
+    def run(self):
+        """ Evaluate the workflow. """
+
```

Comment:
We should probably raise a `NotImplementedError` here for good measure

Reply:
On another issue you suggested just putting a docstring in an empty base class implementation, rather than a NotImplementedError (I think maybe to do with unit testing).

Reply:
Or maybe that was because it was an interface? i.e.
interface = just docstring
base class = return error
Is that the pattern?

Reply:
Yes, that's correct - an `Interface` class expects no implementation, whereas that is not the case for our base classes. If we wanted to we could turn `BaseOperation` into an `ABCHasTraits` class and use abstract methods here, but a more simple way is just to raise a `NotImplementedError` instead